### ISSUE TITLE
Bat search is slow and not accurate sometime

### ISSUE THREAD

**Issue by ciscohack** [sharkdp/bat#1475]
Hi Team,

I am running the latest BAT version and I have noticed search in BAT is slow compare to Less or Vim ( which is superfast in search) Sometimes BAT not even find the search word in the file even though the word exit in searched file.

I request a developer to make BAT tool more efficient and fast in this area

**Comment by j0hnmeow** [comment]
I thought bat uses less for that?

**Comment by ciscohack** [comment]
> I thought bat uses less for that?

Thanks, J0hnmeow, I think you are correct that I believe too but I think we can change the behavior by adding an environment setting in bashrc or zshrc right. Anyway, the problem is when I use Less and search any word it found the word quickly. In Vim it's too fast u just start type and it curser move to the word immediately but in Bat search results are slow and it gets worse when you enabled BAT with -wrap=auto then search failed for the word if it wraps and few words appear in a new line then in that case Bat will never match the search words and throw not found message.  

I believe we can improve it and there is a big scope for improvement. I love if developer address it

**Comment by ciscohack** [comment]
I also love to know if any expert here help me to fix this messing line break and overlap looks weird

![image](https://user-images.githubusercontent.com/11479557/103614855-89551f80-4f4f-11eb-9285-1b599a113e95.png)

**Comment by sharkdp** [comment]
> Thanks, J0hnmeow, I think you are correct that I believe too but I think we can change the behavior by adding an environment setting in bashrc or zshrc right.

Yes. What you are seeing is the search in the pager that is called on your machine (probably `less`, unless configured otherwise - see https://github.com/sharkdp/bat#using-a-different-pager).

Can you please post the output of the script at https://github.com/sharkdp/bat/blob/master/diagnostics/info.sh in order for us to get more information about your system?

> Anyway, the problem is when I use Less and search any word it found the word quickly. In Vim it's too fast u just start type and it curser move to the word immediately but in Bat search results are slow

I can not reproduce this. Search is immediate for me in a 1000+ lines files. How large is your file?

> it gets worse when you enabled BAT with -wrap=auto then search failed for the word if it wraps

That is a (very unfortunate) known problem. Use `--wrap=never` to avoid this. See also #1053 

> and few words appear in a new line then in that case Bat will never match the search words and throw not found message.

I can not reproduce this

> I also love to know if any expert here help me to fix this messing line break and overlap looks weird

Well, this is how it looks like if you disable `bat`s wrapping. Use `--wrap=auto` (which is the default) to avoid this.

**Comment by ciscohack** [comment]
> https://github.com/sharkdp/bat/blob/master/diagnostics/info.sh

Thanks a lot @sharkdp for looking into it. Here is the output of the requested info. I do use -wrap=never but still, the result is the same and looks weird.. with wrap search not yield results sometimes. so I keep it disable but the overall layout result looks bad and need your help to fix if any setting is wrong in my bat config 


system
------

**$ uname -srm**
Darwin 20.2.0 x86_64

**$ sw_vers**
ProductName:    macOS
ProductVersion: 11.1
BuildVersion:   20C69

bat
---

**$ bat --version**
bat 0.17.1

**$ env**
BAT_THEME=gruvbox
BAT_PAGER=less -RJ
PAGER=bat -p

bat_config
----------

**$ cat /Users/praver/.config/bat/config**
```
# This is `bat`s configuration file. Each line either contains a comment or
# a command-line option that you want to pass to `bat` by default. You can
# run `bat --help` to get a list of all possible configuration options.

# Specify desired highlighting theme (e.g. "TwoDark"). Run `bat --list-themes`
# for a list of all available themes
#--theme="Sublime Snazzy"

# Enable this to use italic text on the terminal. This is not supported on all
# terminal emulators (like tmux, by default):
#--italic-text=always

# Uncomment the following line to disable automatic paging:
#--paging=never

# Uncomment the following line if you are using less version >= 551 and want to
# enable mouse scrolling support in `bat` when running inside tmux. This might
# disable text selection, unless you press shift.
#--pager="--RAW-CONTROL-CHARS --quit-if-one-screen --mouse"

# Syntax mappings: map a certain filename pattern to a language.
#   Example 1: use the C++ syntax for .ino files
#   Example 2: Use ".gitignore"-style highlighting for ".ignore" files
#--map-syntax "*.ino:C++"
#--map-syntax ".ignore:Git Ignore"
--map-syntax "*.txt:Python"
--map-syntax "*.log:Python"
```

bat_wrapper
-----------

No wrapper script for 'bat'.

bat_wrapper_function
--------------------

No wrapper function for 'bat'.

**$ /bin/zsh --login -i -c type cat**
```
cat is an alias for bat -l Python --theme=gruvbox --wrap=never
```

tool
----

**$ less --version**
less 487 (POSIX regular expressions)

**Comment by sharkdp** [comment]
>  BAT_PAGER=less -RJ

That probably explains the bad layout. Please try `BAT_PAGER=less` (or leave out the `-J` option and specify `-R` and `-F`)


> **$ /bin/zsh --login -i -c type cat**
> 
> ```
> cat is an alias for bat -l Python --theme=gruvbox --wrap=never
> ```

I think that might be the problem with the slow search. You are telling `bat` to highlight *everything* as Python source code. If you then proceed to show a file with a very different syntax, highlighting instructions will be included everywhere and the output gets very large. Please try to remove `-l Python`.

**Comment by ciscohack** [comment]
> > BAT_PAGER=less -RJ
> 
> That probably explains the bad layout. Please try `BAT_PAGER=less` (or leave out the `-J` option and specify `-R` and `-F`)
> 
> > **$ /bin/zsh --login -i -c type cat**
> > ```
> > cat is an alias for bat -l Python --theme=gruvbox --wrap=never
> > ```
> 
> I think that might be the problem with the slow search. You are telling `bat` to highlight _everything_ as Python source code. If you then proceed to show a file with a very different syntax, highlighting instructions will be included everywhere and the output gets very large. Please try to remove `-l Python`.
> 
> Also, these two lines in the config:
> 
> ```shell
> --map-syntax "*.txt:Python"
> --map-syntax "*.log:Python"
> ```
> 
> tell `bat` to highlight every `*.txt` and every `*.log` file as Python. Is that really what you want?

@sharkdp  I tried setting BAT_PAGER=less -RF but that too not solve the bad layout problem. I found -RJ option in one of your problem ticket so tried to see if that make any difference. but no luck.

yes, I want some keywords to be highlighted so I used this config.  but sometimes it works fast but I can disable to text. but now the main concern is the bad layout and with -wrap=auto layout fixed but then search for a word which wrapped not working.

I tried this option too BAT_PAGER=less with this option layout fix but then line break (wrap not working ) so for readability, i have to scroll horizontally all time which I don't want for a long sentenced line. I want if any long sentence which not fit in terminal size will terminate the new line like the way we do in a wrap.

Could you please help me to achieve it? suggest me setting I will quickly test

**Comment by ciscohack** [comment]
@sharkdp Check this layout after setting BAT_PAGER='less' now layout looks good but complete line is not visible and need horizontal scroll to view the complete line. Can this we adjusted to terminal size long sentence wrap to new line without breaking search functionality 

![image](https://user-images.githubusercontent.com/11479557/103632076-d219d200-4f69-11eb-9956-882a7e9d307e.png)

**Comment by ciscohack** [comment]
@sharkdp Any config suggestion my friend to make layout better with long line break in new line without breaking search functionality

**Comment by sharkdp** [comment]
> Second, I disabled -l python switch in my bat alias to see the search performance very minimal improvement not major.

Please make sure to also remove the `--map-syntax` lines from your config file. Otherwise, .txt files will still be highlighted as Python.

**Comment by ciscohack** [comment]
> > Second, I disabled -l python switch in my bat alias to see the search performance very minimal improvement not major.
> 
> Please make sure to also remove the `--map-syntax` lines from your config file. Otherwise, .txt files will still be highlighted as Python.

@sharkdp Thanks for the update. yes already removed all these lines but I am still not happy with the layout. now the current layout is like I shared in my previous communication where long lines not wrapping and to read the complete line I have to scroll horizontally. can we address this someway. I know wrap is the answer but then this breaks search functionality.

Currently layout :-
https://user-images.githubusercontent.com/11479557/103632076-d219d200-4f69-11eb-9956-882a7e9d307e.png

**Comment by sharkdp** [comment]
@ciscohack Could you please run the diagnostics/info.sh script again with your current setup? Alternatively, could you please try to run the current `master` version of `bat` and use the new `--diagnostic` option?
```
bat --diagnostic
```

You can find up-to-date binaries here: https://github.com/sharkdp/bat/suites/2049050089/artifacts/41269488

**Comment by ciscohack** [comment]
> @ciscohack Could you please run the diagnostics/info.sh script again with your current setup? Alternatively, could you please try to run the current `master` version of `bat` and use the new `--diagnostic` option?
> 
> ```
> bat --diagnostic
> ```
> 
> You can find up-to-date binaries here: https://github.com/sharkdp/bat/suites/2049050089/artifacts/41269488

Hi @sharkdp 

here is the requested output. let me know the next step.

system
------

**$ uname -srm**
Darwin 20.3.0 x86_64

**$ sw_vers**
ProductName:    macOS
ProductVersion: 11.2.1
BuildVersion:   20D74

bat
---

**$ bat --version**
bat 0.17.1

**$ env**
BAT_THEME=gruvbox
BAT_PAGER=less -RF
PAGER=bat

bat_config
----------

**$ cat /Users/praverai/.config/bat/config**
```
# This is `bat`s configuration file. Each line either contains a comment or
# a command-line option that you want to pass to `bat` by default. You can
# run `bat --help` to get a list of all possible configuration options.

# Specify desired highlighting theme (e.g. "TwoDark"). Run `bat --list-themes`
# for a list of all available themes
#--theme="Sublime Snazzy"

# Enable this to use italic text on the terminal. This is not supported on all
# terminal emulators (like tmux, by default):
#--italic-text=always

# Uncomment the following line to disable automatic paging:
#--paging=never

# Uncomment the following line if you are using less version >= 551 and want to
# enable mouse scrolling support in `bat` when running inside tmux. This might
# disable text selection, unless you press shift.
#--pager="--RAW-CONTROL-CHARS --quit-if-one-screen --mouse"

# Syntax mappings: map a certain filename pattern to a language.
#   Example 1: use the C++ syntax for .ino files
#   Example 2: Use ".gitignore"-style highlighting for ".ignore" files
#--map-syntax "*.ino:C++"
#--map-syntax ".ignore:Git Ignore"
#--map-syntax "*.txt:Python"
#--map-syntax "*.log:Python"
```

bat_wrapper
-----------

No wrapper script for 'bat'.

bat_wrapper_function
--------------------

No wrapper function for 'bat'.

**$ /bin/zsh --login -i -c type cat**
```
cat is an alias for bat --terminal-width=-2 --style full --theme=gruvbox --wrap=never --decorations=always --color=always
```

tool
----

**$ less --version**
less 487 (POSIX regular expressions)

**Comment by sharkdp** [comment]
You have a lot of settings in your `cat` alias.

> --terminal-width=-2

Could you please try to remove that? And maybe the other settings as well?

**Comment by ciscohack** [comment]
> You have a lot of settings in your `cat` alias.
> 
> > --terminal-width=-2
> 
> Could you please try to remove that? And maybe the other settings as well?

You want me to simplify this config right 

alias cat='bat --terminal-width=-2 --style full --theme=gruvbox --wrap=never --decorations=always --color=always'

I have changed the above config to 

alias cat='bat'

is this okay now but the layout still not good I mean wrap break search functionality and pager less use no-wrap force me to scroll horizontal to real long sentence

**Comment by sharkdp** [comment]
I am sorry but this is very hard to follow.



> I mean wrap break search functionality

That is unfortunate, but expected. It's not going to be solved easily. See #1053 for details.

> and pager less use no-wrap force me to scroll horizontal to real long sentence

?

**Comment by ciscohack** [comment]
> I am sorry but this is very hard to follow.
> 
> > I mean wrap break search functionality
> 
> That is unfortunate, but expected. It's not going to be solved easily. See #1053 for details.
> 
> > and pager less use no-wrap force me to scroll horizontal to real long sentence
> 
> ?

Okay thanks, I tried many things but no success so then now i will leave the way it's rendering as we can't do much

**Comment by sharkdp** [comment]
Ok. Let us know if you think there is anything we can actually do.
